flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-iceberg/src/main/java/org/apache/flink/cdc/connectors/iceberg/sink/v2/compaction/CompactionOperator.java

```diff
@@ -0,0 +1,154 @@
+import org.apache.flink.util.concurrent.ExecutorThreadFactory;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.CatalogUtil;
+import org.apache.iceberg.actions.RewriteDataFilesActionResult;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.flink.actions.Actions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/** An Operator to add checkpointId to IcebergTable Small file compaction. */
+public class CompactionOperator
+        extends AbstractStreamOperator<CommittableMessage<WriteResultWrapper>>
+        implements OneInputStreamOperator<
+                CommittableMessage<WriteResultWrapper>, CommittableMessage<WriteResultWrapper>> {
+    protected static final Logger LOGGER = LoggerFactory.getLogger(CompactionOperator.class);
+
+    private final Map<String, String> catalogOptions;
+
+    private Catalog catalog;
+
+    /** store a list of MultiTableCommittable in one checkpoint. */
+    private final Map<TableId, Integer> tableCommitTimes;
+
+    private final Set<TableId> compactedTables;
+
+    private final CompactionOptions compactionOptions;
+
+    private volatile Throwable throwable;
+
+    private ExecutorService compactExecutor;
+
+    public CompactionOperator(
+            Map<String, String> catalogOptions, CompactionOptions compactionOptions) {
+        this.tableCommitTimes = new HashMap<>();
+        this.compactedTables = new HashSet<>();
+        this.catalogOptions = catalogOptions;
+        this.compactionOptions = compactionOptions;
+    }
+
+    @Override
+    public void open() throws Exception {
+        super.open();
+        if (compactExecutor == null) {
+            this.compactExecutor =
+                    Executors.newSingleThreadScheduledExecutor(
+                            new ExecutorThreadFactory(
+                                    Thread.currentThread().getName() + "-Cdc-Compaction"));
+        }
+    }
+
+    @Override
+    public void processElement(StreamRecord<CommittableMessage<WriteResultWrapper>> element) {
+        if (element.getValue() instanceof CommittableWithLineage) {
+            TableId tableId =
+                    ((CommittableWithLineage<WriteResultWrapper>) element.getValue())
+                            .getCommittable()
+                            .getTableId();
+            tableCommitTimes.put(tableId, tableCommitTimes.getOrDefault(tableId, 0) + 1);
+            int commitTimes = tableCommitTimes.get(tableId);
+            if (commitTimes >= compactionOptions.getCommitInterval()
+                    && !compactedTables.contains(tableId)) {
+                if (throwable != null) {
+                    throw new RuntimeException(throwable);
```

Comment:
Do we need to move this if statement to the start of this method `processElement` out of `if (element.getValue() instanceof CommittableWithLineage)`?